File: 03-hard/0044-following-integer/0044-following-integer.py
#!/usr/bin/env python
# -*- coding: utf-8; py-indent-offset:4 -*-
###############################################################################
#
# Copyright (C) 2015 Daniel Rodriguez
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################
from __future__ import (absolute_import, division, print_function,)
#                        unicode_literals)

import bisect
import itertools
import sys

# Small Py2/3 compatibility layer
if sys.version_info.major == 2:
    MAXINT = sys.maxint
    MININT = -sys.maxint - 1

    filter = itertools.ifilter
    map = itertools.imap
    range = xrange
    zip = itertools.izip

else:  # >= 3
    MAXINT = sys.maxsize
    MININT = -sys.maxsize - 1

# This solution avoids using the built-in itertools.permutations and set and
# ... which can be memory/cpu intensive if the amount of digits starts growing


def following_integer(n, ZERO='0'):
    # find out if moving one digit forward gives us the next
    # 115 -> 151, 2048 -> 2084, 20005, 20050
    ln = len(n)
    # loop backwards to compare tail to predecessor
    for i in range(ln - 1, 0, -1):
        ni1 = n[i - 1]
        # find any number greater in range greater than predecessor
        gti = [(x, j) for j, x in enumerate(n[i:]) if x > ni1]
        if not gti:
            continue

        # swap predecessor, against the minimum of the greater
        gtidx = min(gti)[1] + i  # compensate enumerate 0 based j
        n[i - 1], n[gtidx] = n[gtidx], n[i - 1]

        # return up to and including predecessor position + minimum of next
        # range (sorted)
        return n[0:i] + sorted(n[i:])

    # Need to grow digit length. Zero inserted as 2nd leftmost digit
    # After pushing the lowest greater than 0 digit to the left
    # 842 -> 248 -> 2048, 8420 -> 0248 -> 20048
    sn = sorted(n)
    # get first which is not zero
    ifirst = next(i for i in range(ln) if sn[i] != ZERO)
    # return the new string with a 2nd leftmost zero
    n = [sn.pop(ifirst), ZERO] + sn

    return n


if __name__ == '__main__':
    test_cases = open(sys.argv[1], 'r')
    for line in test_cases:
        n = list(line.rstrip('\n'))
        o = following_integer(n)
        print(''.join(o))

    test_cases.close()

File: 03-hard/0044-following-integer/0044-following-integer.cpp
/*

  Copyright (C) 2015 Daniel Rodriguez

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

// Headers for test case
#include <fstream>
#include <iostream>

// Headers for the implementation
#include <algorithm>
#include <iterator>
#include <vector>

template <typename InT, typename OutT>
auto
following_integer(InT first, InT last, OutT out)
{
    using InType = typename std::iterator_traits<InT>::value_type;
    auto rlast = std::reverse_iterator<InT>(last);
    auto rfirst = std::reverse_iterator<InT>(first);

    // in the 1st part we try to move one digit forward to make the figure
    // greater and then sort the lower part to have the mininum possible value
    for(auto start=rlast, target=std::next(rlast); target != rfirst; start++, target++) {
        auto mingreat = std::numeric_limits<InType>::max();
        auto minit = target;

        // find the min greater than target, to do a swap
        for(auto check=rlast; check != target; check++) {
            if(*check > *target and *check < mingreat) {
                minit = check;
                mingreat = *check;
            }
        }

        if(minit == target)
            continue;  // no min greater continue search

        std::iter_swap(minit, target);  // found: swap values

        // find out how far away from the beginning we stopped
        auto d = std::distance(target, rfirst);
        std::sort(std::next(first, d), last);  // sort only the lower part
        std::copy(first, last, out);  // send to output iterator
        return;
    }

    // no movement of digits is possible to make it larger. Need to get lowest
    // ordering and insert a 2nd left most zero. The lowest ordering is the one
    // with a significant digit (non zero) at the left most part

    std::sort(first, last);  // lower value
    // find first nonzero
    auto nonzero = std::find_if(first, last, [](const InType &a) { return a;});
    std::iter_swap(first, nonzero);  // swap values

    *out++ = *first;  // output now lower
    *out++ = not bool(*first);  // output the zero, calculate it
    std::copy(std::next(first), last, out);  // copy the rest out
}

auto
int_range(int number)
{
    // receive an int an return and return a range with the digits
    const auto MAX_POSITIONS = 6;
    static std::vector<int> vnumber(MAX_POSITIONS);  // n < 10^6
    auto vbegin = vnumber.begin();

    auto i = 0;
    while(number) {
        vnumber[i++] = number % 10;
        number /= 10;
    }

    auto vend = std::next(vbegin, i);
    std::reverse(vbegin, vend);
    return std::make_pair(vbegin, vend);
}


int
main(int argc, char *argv[]) {
    std::ifstream stream(argv[1]);

    auto out = std::ostream_iterator<int>(std::cout);

    int number;
    while(stream >> number) {
        auto range = int_range(number);
        following_integer(range.first, range.second, out);
        std::cout << std::endl;

    }
    return 0;
}
